clash-lib/src/Clash/Normalize/Util.hs

```diff
@@ -67,10 +69,47 @@ import           Clash.Rewrite.Types
   (RewriteMonad, TransformContext(..), bindings, curFun, dbgLevel, extra,
    tcCache)
 import           Clash.Rewrite.Util
-  (runRewrite, specialise, mkTmBinderFor, mkDerivedName)
+  (runRewrite, specialise, mkTmBinderFor, mkDerivedName, isConstant)
 import           Clash.Unique
 import           Clash.Util
-  (SrcSpan, anyM, makeCachedU, traceIf, mapAccumLM)
+  (SrcSpan, allM, anyM, makeCachedU, traceIf, mapAccumLM)
+
+-- | Determine whether a term does any work, i.e. adds to the size of the circuit
+isWorkFree
+  :: Term
+  -> NormalizeSession Bool
+isWorkFree (collectArgs -> (fun,args)) = case fun of
+  Var i -> do
+    let isPoly = isPolyFunTy (varType i)
+    if isLocalId i && not isPoly then
+      pure True
+    else
+      not <$> isRecursiveBndr i
```

Comment:
We should check whether the global function is work-free

Reply:
Done?